cpp/oneapi/dal/table/detail/access_iface.hpp

```diff
@@ -21,81 +21,275 @@
-using v1::access_iface_dpc;
+    DECLARE_PUSH_COLUMN_DPC(float)
+    DECLARE_PUSH_COLUMN_DPC(double)
+    DECLARE_PUSH_COLUMN_DPC(std::int32_t)
 #endif
+};
+
+template <typename Derived>
+class push_column_template : public base, public push_column_iface {
+public:
+    DEFINE_TEMPLATE_PUSH_COLUMN_HOST(float)
+    DEFINE_TEMPLATE_PUSH_COLUMN_HOST(double)
+    DEFINE_TEMPLATE_PUSH_COLUMN_HOST(std::int32_t)
+
+#ifdef ONEDAL_DATA_PARALLEL
+    DEFINE_TEMPLATE_PUSH_COLUMN_DPC(float)
+    DEFINE_TEMPLATE_PUSH_COLUMN_DPC(double)
+    DEFINE_TEMPLATE_PUSH_COLUMN_DPC(std::int32_t)
+#endif
+};
+
+#undef PULL_ROWS_SIGNATURE_HOST
+#undef PULL_ROWS_SIGNATURE_DPC
+#undef PULL_COLUMN_SIGNATURE_HOST
+#undef PULL_COLUMN_SIGNATURE_DPC
+#undef PUSH_ROWS_SIGNATURE_HOST
+#undef PUSH_ROWS_SIGNATURE_DPC
+#undef PUSH_COLUMN_SIGNATURE_HOST
+#undef PUSH_COLUMN_SIGNATURE_DPC
+#undef DECLARE_PULL_ROWS_HOST
+#undef DECLARE_PULL_ROWS_DPC
+#undef DECLARE_PULL_COLUMN_HOST
+#undef DECLARE_PULL_COLUMN_DPC
+#undef DECLARE_PUSH_ROWS_HOST
+#undef DECLARE_PUSH_ROWS_DPC
+#undef DECLARE_PUSH_COLUMN_HOST
+#undef DECLARE_PUSH_COLUMN_DPC
+#undef DEFINE_TEMPLATE_PULL_ROWS_HOST
+#undef DEFINE_TEMPLATE_PULL_ROWS_DPC
+#undef DEFINE_TEMPLATE_PULL_COLUMN_HOST
+#undef DEFINE_TEMPLATE_PULL_COLUMN_DPC
+#undef DEFINE_TEMPLATE_PUSH_ROWS_HOST
+#undef DEFINE_TEMPLATE_PUSH_ROWS_DPC
+#undef DEFINE_TEMPLATE_PUSH_COLUMN_HOST
+#undef DEFINE_TEMPLATE_PUSH_COLUMN_DPC
+
+template <typename Object>
+inline std::shared_ptr<pull_rows_iface> get_pull_rows_iface(Object&& obj) {
+    const auto pimpl = pimpl_accessor{}.get_pimpl(std::forward<Object>(obj));
+    return std::shared_ptr<pull_rows_iface>{ pimpl, pimpl->get_pull_rows_iface() };
+}
+
+template <typename Object>
+inline std::shared_ptr<pull_column_iface> get_pull_column_iface(Object&& obj) {
+    const auto pimpl = pimpl_accessor{}.get_pimpl(std::forward<Object>(obj));
+    return std::shared_ptr<pull_column_iface>{ pimpl, pimpl->get_pull_column_iface() };
+}
+
+template <typename Object>
+inline std::shared_ptr<push_rows_iface> get_push_rows_iface(Object&& obj) {
+    const auto pimpl = pimpl_accessor{}.get_pimpl(std::forward<Object>(obj));
+    return std::shared_ptr<push_rows_iface>{ pimpl, pimpl->get_push_rows_iface() };
+}
+
+template <typename Object>
+inline std::shared_ptr<push_column_iface> get_push_column_iface(Object&& obj) {
+    const auto pimpl = pimpl_accessor{}.get_pimpl(std::forward<Object>(obj));
+    return std::shared_ptr<push_column_iface>{ pimpl, pimpl->get_push_column_iface() };
+}
+
+} // namespace v1
+
+using v1::pull_rows_iface;
+using v1::pull_rows_template;
+using v1::pull_column_iface;
+using v1::pull_column_template;
+using v1::push_rows_iface;
+using v1::push_rows_template;
+using v1::push_column_iface;
+using v1::push_column_template;
+using v1::get_pull_rows_iface;
+using v1::get_pull_column_iface;
+using v1::get_push_rows_iface;
+using v1::get_push_column_iface;
 
 } // namespace oneapi::dal::detail
```

Comment:
Maybe to split this file into several files that contain just a single interface declaration?